### ISSUE TITLE
Will responds to all messages in room when initially invited. Stops after restart.

### ISSUE THREAD

**Issue by netjunki** [skoczen/will#346]
If you invite will to a room they respond to all messages whether directed to will or not.

<img width="435" alt="will_new_channel" src="https://user-images.githubusercontent.com/64289/38449977-f4549f2e-39ca-11e8-8877-e025f1190c0a.PNG">

I've only been seeing this in slack. A restart of the bot python will correct the issue.

**Comment by Ashex** [comment]
@netjunki Did you find a solution to this? I'm getting ready to dig into the slack backend to see how this is happening.

**Comment by netjunki** [comment]
I still haven't. We only add the bot to a channel about once a month nowadays so it hasn't come up in a bit. But when it does... well it can be hilarious. :-) I suspect the issue has something to do with will not updating the channel list metadata when it's asked to join a channel... maybe it's not consuming some slack "channel join" even correctly? But this doesn't seem to throw an exception so need to work backwards from where it publishes those notifications to figure out where the problem is.

**Comment by Ashex** [comment]
There is a race condition with the channel update (#354) when will joins a new channel but an exception is only thrown when it's a private channel. 

I didn't see it when I was using a slack token associated with a user since Will is aware of when it joins/leaves a channel.

I started noticing it when I switched over to a bot token which made will a Slack app. The crummy part about Apps is that you have zero awareness of what channel you're in so will sort of starts receiving commands from a channel it didn't know it was in and for some reason this is confusing it. Technically it can listen in on the RTM feed to find when the app is added but I think that's a little too intensive.

**Comment by skoczen** [comment]
Quick note - seeing all this, expect to do a big catch-up on Will next week.   Moving to Paris on Tuesday, and then will be in catch-up mode then.

In good news, I have a new gig where we'll be using will for mission critical devops.  Yay dogfooding!

More soon!
-Steven

**Comment by Ashex** [comment]
@skoczen Welcome back! 

I believe I've isolated the issue. There's a bug somewhere that causes this check:

```python
            if len(channel.members.keys()) == 0:
                is_private_chat = True
```

To evaluate to true when the bot joins a new channel due to the members array being empty. I haven't really figured out why members is empty on join, I added an extra check earlier to update channels when the channel is missing so it is fresh. 
I worked around this by changing the check to the following since we know that the channel id will always be different from the channel name but these values are the same for direct messages.

```python
            if channel.id == channel.name:
                is_private_chat = True
```
